### PR TITLE
Introduce single point of interaction

### PR DESCRIPTION
**Pull request by 34x4p08**
In this protocol update:

- Single contract to manage all the CDPs
- Onchain CDP registry
- OracleType-based logic of CDP management
- Refactor oracle contract for LP tokens
- Refactor of Chainlink-based oracle contract
- 0.7.6 compiler version

### CODE REVIEW

File: contracts/vault-managers/CDPManager01.sol [hunk 1]
```diff
@@ -0,0 +1,355 @@
...
+                // check oracle
+                require(oracleRegistry.oracleByAsset(asset) != address(0), "Unit Protocol: DISABLED_ORACLE");
+
+                // withdraw collateral to the user address
+                vault.withdrawMain(asset, msg.sender, assetAmount);
+
+                if (usdpAmount != 0) {
+                    uint fee = vault.calculateFee(asset, msg.sender, usdpAmount);
+                    vault.chargeFee(vault.usdp(), msg.sender, fee);
+                    vault.repay(asset, msg.sender, usdpAmount);
```
Comment (Eenae) [inline contracts/vault-managers/CDPManager01.sol:170]: `_repay`?

File: contracts/vault-managers/CDPManager01.sol [hunk 2]
```diff
@@ -0,0 +1,355 @@
...
+        uint debtAfter = vault.repay(asset, user, usdpAmount);
+        if (debtAfter == 0) {
+            // clear unused storage
+            vault.destroy(asset, user);
+        }
+    }
+
+    function _ensurePositionCollateralization(address asset, address user) internal view {
+        // collateral value of the position in USD
+        uint usdValue_q112 = IOracleUsd(oracleRegistry.oracleByAsset(asset)).assetToUsd(asset, vault.collaterals(asset, user));
```
Comment (Eenae) [inline contracts/vault-managers/CDPManager01.sol:239]: We're ignoring `vault.oracleType[asset][user]`. Is it ok?

Comment (Eenae) [reply contracts/vault-managers/CDPManager01.sol:239]: Moreover, we're ignoring `vault.isOracleTypeEnabled` as well. However, this setting may block liquidations, so should be in sync with borrowing.

File: contracts/vault-managers/CDPManager01.sol [hunk 3]
```diff
@@ -0,0 +1,355 @@
...
+        address asset,
+        address user
+    ) public view returns (uint) {
+        uint debt = vault.getTotalDebt(asset, user);
+        if (debt == 0) return uint(-1);
+        
+        uint usdValue_q112 = IOracleUsd(oracleRegistry.oracleByAsset(asset)).assetToUsd(asset, vault.collaterals(asset, user));
+
+        return debt.mul(100).mul(Q112).div(usdValue_q112);
+    }
```
Comment (Eenae) [inline contracts/vault-managers/CDPManager01.sol:328]: Copy-paste should be avoided.
`utilizationRatio` of zero debt is zero, isn't it?

File: contracts/vault-managers/CDPManager01.sol [hunk 4]
```diff
@@ -0,0 +1,355 @@
...
+    function liquidationPrice_q112(
+        address asset,
+        address user
+    ) external view returns (uint) {
+        uint debt = vault.getTotalDebt(asset, user);
+        if (debt == 0) return uint(-1);
+        
+        uint collateralLiqPrice = debt.mul(100).mul(Q112).div(vaultManagerParameters.liquidationRatio(asset));
+        
+        require(IToken(asset).decimals() <= 18, "Unit Protocol: NOT_SUPPORTED_DECIMALS");
```
Comment (Eenae) [inline contracts/vault-managers/CDPManager01.sol:346]: Now it's too late. Must be checked on `join`.

File: contracts/interfaces/IOracleUsd.sol [hunk 5]
```diff
@@ -0,0 +1,5 @@
+interface IOracleUsd {
+
+    // returns Q112-encoded value
+    function assetToUsd(address asset, uint amount) external view returns (uint);
```
[the remark this answers is not in the document]
Comment (34x4p08) [reply contracts/interfaces/IOracleUsd.sol:4]: Yes

File: contracts/interfaces/IOracleEth.sol [hunk 6]
```diff
@@ -0,0 +1,11 @@
+interface IOracleEth {
+
+    // returns Q112-encoded value
+    function assetToEth(address asset, uint amount) external view returns (uint);
```
[the remark this answers is not in the document]
Comment (34x4p08) [reply contracts/interfaces/IOracleEth.sol:4]: Yes

File: contracts/vault-managers/CDPManager01.sol (hunk 2 above)
Comment (34x4p08) [reply contracts/vault-managers/CDPManager01.sol:239]: Oracle registry should be in sync with `isOracleTypeEnabled`

**Review by Eenae — CHANGES_REQUESTED** [review]

File: contracts/interfaces/IOracleUsd.sol (hunk 5 above)
[the remark this answers is not in the document]
Comment (Eenae) [reply contracts/interfaces/IOracleUsd.sol:4]: It's worth adding such a comment

File: contracts/interfaces/IOracleEth.sol (hunk 6 above)
[the remark this answers is not in the document]
Comment (Eenae) [reply contracts/interfaces/IOracleEth.sol:4]: It's worth adding such a comment

File: contracts/oracles/OracleRegistry.sol [hunk 7]
```diff
@@ -3,33 +3,79 @@
...
-     * @param asset The address of the collateral
-     * @param oracle The oracle address
-     * @param oracleType The oracle type ID
-     **/
-    function setOracle(address asset, address oracle, uint oracleType) public onlyManager {
+    function setOracleTypeToAsset(address asset, uint oracleType) public onlyManager {
         require(asset != address(0) && oracleType != 0, "Unit Protocol: INVALID_ARGS");
-        oracleByAsset[asset] = oracle;
+        oracleTypeByAsset[asset] = oracleType;
+        emit AssetOracle(asset, oracleType);
```
[the remark this answers is not in the document]
Comment (34x4p08) [reply contracts/oracles/OracleRegistry.sol:46]: `maxOracleType` is updated in `setOracle`

File: contracts/oracles/OracleRegistry.sol [hunk 8]
```diff
@@ -3,33 +3,153 @@
...
-     * @param oracle The oracle address
-     * @param oracleType The oracle type ID
-     **/
-    function setOracle(address asset, address oracle, uint oracleType) public onlyManager {
-        require(asset != address(0) && oracleType != 0, "Unit Protocol: INVALID_ARGS");
-        oracleByAsset[asset] = oracle;
+    function setKeydonixOracleTypes(uint[] calldata _keydonixOracleTypes) public onlyManager {
+        while (keydonixOracleTypes.length != 0) {
+            keydonixOracleTypes.pop();
+        }
```
[the remark this answers is not in the document]
Comment (34x4p08) [reply contracts/oracles/OracleRegistry.sol:59]: What about gas refund in this case?

File: contracts/oracles/OracleRegistry.sol (hunk 7 above)
[the remark this answers is not in the document]
Comment (Eenae) [reply contracts/oracles/OracleRegistry.sol]: ok, why shouldn't we call it here as well?
Alternatively, if it's expected that these external calls are somehow always consistent, we can `assert` that `maxOracleType` was updated properly.

[the remark this answers is not in the document]
Comment (Eenae) [reply contracts/oracles/OracleRegistry.sol]: ok, resolved by `validAddress(oracleByType[oracleType])`

**Review by Eenae — APPROVED** [review]

**Pull request merged**